### ISSUE TITLE
js/Ice/ami failure (Windows)

### ISSUE THREAD

**Issue by bernardnormier** [zeroc-ice/ice#2979]
From https://github.com/zeroc-ice/ice/actions/runs/11542801641/job/32126434008?pr=2978

```
*** [209/270] Running js/Ice/ami tests ***
[ running client/server test - 10/27/24 19:19:02 ]
- Config: tcp,x64
(D:\a\ice\ice\cpp\test\Ice\ami\msbuild\server\x64\Release\server.exe --Ice.Default.Host=127.0.0.1 --Test.BasePort=14100 --Ice.Warn.Connections=1 --Ice.Default.Protocol=tcp --Ice.PrintStackTraces=1 --Ice.ThreadPool.Server.Size=1 --Ice.ThreadPool.Server.SizeMax=3 --Ice.ThreadPool.Server.SizeWarn=0 --Ice.PrintAdapterReady=1 --Ice.Trace.Network=2 --Ice.Trace.Retry=1 --Ice.Trace.Protocol=1 --Ice.LogFile=D:\a\ice\ice\js\test\Ice\ami\server-102724-1919.log env={'PATH': 'D:\\a\\ice\\ice\\cpp\\test\\Common\\msbuild\\x64\\Release;D:\\a\\ice\\ice\\cpp\\bin\\x64\\Release'})
(node D:\a\ice\ice\js\test\Common/run.js file://D:\a\ice\ice\js\test\Ice/ami\Client.js Client --Ice.Default.Host=127.0.0.1 --Test.BasePort=14100 --Ice.Warn.Connections=1 --Ice.Default.Protocol=tcp)
testing batch requests with proxy... ok
testing batch requests with connection... ok
testing batch requests with communicator... ok
testing AsyncResult operations... ok
testing bi-dir... ok
file:///D:/a/ice/ice/js/node_modules/ice/src/Ice/ConnectionI.js:188
        this.setState(StateClosing, new CommunicatorDestroyedException());
                                    ^

Ice.CommunicatorDestroyedException: Communicator destroyed.
    at ConnectionI.destroy (file:///D:/a/ice/ice/js/node_modules/ice/src/Ice/ConnectionI.js:188:37)
    at file:///D:/a/ice/ice/js/node_modules/ice/src/Ice/OutgoingConnectionFactory.js:35:70
    at Array.forEach (<anonymous>)
    at ConnectionListMap.forEach (file:///D:/a/ice/ice/js/node_modules/ice/src/Ice/OutgoingConnectionFactory.js:509:25)
    at OutgoingConnectionFactory.destroy (file:///D:/a/ice/ice/js/node_modules/ice/src/Ice/OutgoingConnectionFactory.js:35:37)
    at Instance.destroy (file:///D:/a/ice/ice/js/node_modules/ice/src/Ice/InstanceExtensions.js:333:45)
    at Communicator.destroy (file:///D:/a/ice/ice/js/node_modules/ice/src/Ice/Communicator.js:23:31)
    at Client.run (file:///D:/a/ice/ice/js/test/Ice/ami/Client.js:270:36)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async file:///D:/a/ice/ice/js/test/Common/run.js:28:9

Node.js v20.18.0
saved D:\a\ice\ice\js\test\Ice\ami\server-102724-1919.log
```

**Comment by pepone** [comment]
I was unable to reproduce this but fixed a number of issues that were affecting communicator destruction:

#2993 
#3001 
#3046

**Comment by bernardnormier** [comment]
It's still failing today:
https://github.com/zeroc-ice/ice/actions/runs/11692290200/job/32561315719

```
*** [211/271] Running js/Ice/ami tests ***
[ running client/server test - 11/05/24 21:51:44 ]
- Config: tcp,x64
(D:\a\ice\ice\cpp\test\Ice\ami\msbuild\server\x64\Release\server.exe --Ice.Default.Host=127.0.0.1 --Test.BasePort=14100 --Ice.Warn.Connections=1 --Ice.Default.Protocol=tcp --Ice.PrintStackTraces=1 --Ice.ThreadPool.Server.Size=1 --Ice.ThreadPool.Server.SizeMax=3 --Ice.ThreadPool.Server.SizeWarn=0 --Ice.PrintAdapterReady=1 --Ice.Trace.Network=2 --Ice.Trace.Retry=1 --Ice.Trace.Protocol=1 --Ice.LogFile=D:\a\ice\ice\js\test\Ice\ami\server-110524-2151.log env={'PATH': 'D:\\a\\ice\\ice\\cpp\\test\\Common\\msbuild\\x64\\Release;D:\\a\\ice\\ice\\cpp\\bin\\x64\\Release'})
(node D:\a\ice\ice\js\test\Common/run.js file://D:\a\ice\ice\js\test\Ice/ami\Client.js Client --Ice.Default.Host=127.0.0.1 --Test.BasePort=14100 --Ice.Warn.Connections=1 --Ice.Default.Protocol=tcp)
testing batch requests with proxy... ok
testing batch requests with connection... ok
testing batch requests with communicator... ok
testing AsyncResult operations... ok
testing bi-dir... ok
file:///D:/a/ice/ice/js/node_modules/ice/src/Ice/ConnectionI.js:188
        this.setState(StateClosing, new CommunicatorDestroyedException());
                                    ^

Ice.CommunicatorDestroyedException: Communicator destroyed.
    at ConnectionI.destroy (file:///D:/a/ice/ice/js/node_modules/ice/src/Ice/ConnectionI.js:188:37)
    at file:///D:/a/ice/ice/js/node_modules/ice/src/Ice/OutgoingConnectionFactory.js:35:70
    at Array.forEach (<anonymous>)
    at ConnectionListMap.forEach (file:///D:/a/ice/ice/js/node_modules/ice/src/Ice/OutgoingConnectionFactory.js:475:25)
    at OutgoingConnectionFactory.destroy (file:///D:/a/ice/ice/js/node_modules/ice/src/Ice/OutgoingConnectionFactory.js:35:37)
    at Instance.destroy (file:///D:/a/ice/ice/js/node_modules/ice/src/Ice/InstanceExtensions.js:338:45)
    at Communicator.destroy (file:///D:/a/ice/ice/js/node_modules/ice/src/Ice/Communicator.js:23:31)
    at Client.run (file:///D:/a/ice/ice/js/test/Ice/ami/Client.js:278:36)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async file:///D:/a/ice/ice/js/test/Common/run.js:28:9

Node.js v20.18.0
saved D:\a\ice\ice\js\test\Ice\ami\server-110524-2151.log

test in Ice/ami failed:

```

**Comment by pepone** [comment]
I was able to reproduce by running on a local Windows VM with 2 cores, while running cpu stress in the host.

I would try with some additional trace to understand how this happens.

**Comment by pepone** [comment]
I added more tracing and seems like the problem is that one of the opWithPayload calls is getting and unhandled rejection.

There is two opWithPayload calls in a loop, which are intended to fill the send buffers until the transport blocks.

The test just awaits the last call and assumes that all other calls completed before it. Which seems like a correct assumption.

The rejection shows the opWithPayload is rejected with the `CommunicatorDestroyedException` from the bug report. Need to further debug to see why this happens.

**Comment by pepone** [comment]
Fixed in #3097
